common/flatpak-dir.c

```diff
@@ -11624,6 +11624,91 @@ flatpak_dir_cleanup_remote_for_url_change (FlatpakDir   *self,
   return TRUE;
 }
 
+static GBytes *
+load_keyring_for_remote (FlatpakDir    *self,
+                         const char    *remote_name,
+                         GCancellable  *cancellable,
+                         GError       **error)
+{
+  g_autofree char *keyring_name = NULL;
+  g_autoptr(GFile) keyring_file = NULL;
+  gchar *contents;
+  gsize len;
+
+  keyring_name = g_strdup_printf ("%s.trustedkeys.gpg", remote_name);
+  keyring_file = g_file_get_child (ostree_repo_get_path (self->repo), keyring_name);
+
+  /* This might fail if the remote is using a global keyrings directory which
+   * should be handled gracefully */
+  if (g_file_load_contents (keyring_file, cancellable, &contents, &len, NULL, error))
+    return g_bytes_new_take (g_steal_pointer (&contents), len);
+
+  return NULL;
+}
+
+static gboolean
+check_for_collection_collisions (FlatpakDir    *self,
+                                 const char    *remote_name,
+                                 const char    *collection_id,
+                                 GBytes        *gpg_data,
+                                 GCancellable  *cancellable,
+                                 GError       **error)
+{
+  g_autoptr(GBytes) keyring_bytes = NULL;
+  g_autoptr(GError) local_error = NULL;
+
+  if (gpg_data != NULL)
+    keyring_bytes = g_bytes_ref (gpg_data);
+  else
+    keyring_bytes = load_keyring_for_remote (self, remote_name, cancellable, &local_error);
+
+  /* Treat errors loading the keyring as non-fatal in case the remote uses
+   * a global keyrings directory */
+  if (keyring_bytes == NULL)
+    {
+      g_debug ("%s: Failed to load keyring for remote %s: %s",
+               G_STRFUNC, remote_name, local_error->message);
+    }
+  else
+    {
+      g_auto(GStrv) remotes = NULL;
+      int i;
+
+      remotes = flatpak_dir_list_remotes (self, cancellable, error);
+      if (remotes == NULL)
+        return FALSE;
+
+      for (i = 0; remotes[i] != NULL; i++)
+        {
+          g_autofree char *other_remote_collection_id = NULL;
+          g_autoptr(GBytes) other_keyring_bytes = NULL;
+
+          if (g_strcmp0 (remote_name, remotes[i]) == 0)
+            continue;
+
+          other_remote_collection_id = flatpak_dir_get_remote_collection_id (self, remotes[i]);
+          if (g_strcmp0 (collection_id, other_remote_collection_id) != 0)
+            continue;
+
+          other_keyring_bytes = load_keyring_for_remote (self, remotes[i], cancellable, &local_error);
+          if (other_keyring_bytes == NULL)
+            {
+              g_debug ("%s: Failed to load keyring for remote %s: %s",
+                       G_STRFUNC, remotes[i], local_error->message);
+              g_clear_error (&local_error);
+              continue;
+            }
+
+          if (!g_bytes_equal (keyring_bytes, other_keyring_bytes))
```

Comment:
I worry about this check. The keyring doesn't have to be identical for the gpg keys to match. For example any of the keyrings could contain two keys, or there might be some difference in version in the gpg keyring format or other gratious difference.

Reply:
Also, how does this play with key updates (via metadata). Do those always end up with identical copeis of the keyring?

Reply:
> Also, how does this play with key updates (via metadata). Do those always end up with identical copeis of the keyring?

This check doesn't happen for a keyring update because we don't compare a remote to its own config


> I worry about this check. The keyring doesn't have to be identical for the gpg keys to match. For example any of the keyrings could contain two keys, or there might be some difference in version in the gpg keyring format or other gratious difference.

I don't think we have to worry about the keyrings having different sets of keys, because a mirror would use the same set of keys. But I'm not sure what can be done about the file format differing. We could potentially try to verify some signatures (say, on `ostree-metadata`) that are valid for the existing remote and see if they're valid for the one being modified, but then we have to assume we have some signed data, or do network I/O here to get it.

Reply:
On the other hand, why would a mirror have the same keys in a different format?

Reply:
> > Also, how does this play with key updates (via metadata). Do those always end up with identical copeis of the keyring?
> 
> This check doesn't happen for a keyring update because we don't compare a remote to its own config

Sure, but you have a flathub remote, with gpg key 1, now it rotates to gpg key 2, importing it into the local keyring. Now your local gpg key list is updated, and may not look bitwise identical to "upstream gpg key 2". At this point a third party repo may claim to use the same gpg key as the "upstream gpg key 2" and we will not match this as bad.

Reply:
If we were to continue down this path I think the way to do it would be to use gpgme and specifically the `keyid` attribute documented [here](https://www.gnupg.org/documentation/manuals/gpgme/Key-objects.html#Key-objects), and see if the remote being modified would have a (possibly empty) subset of the keys associated with the existing remote with the same collection ID (which would be an error) or the same set of keys (which would not be an error).

However I think that would break keyring updates (via `xa.gpg-keys`) in the case where you have a mirror remote configured (say you have both `flathub` and `flathub-mirror`), because you wouldn't update both in one atomic transaction. From the perspective of `flatpak_dir_modify_remote()` it would look like you're changing the set of keys associated with `flathub` to differ from the set associated with `flathub-mirror` (even though in reality that would only be the case until `flathub-mirror` gets the same update).

Therefore I'll try the strategy described in the last sentence [here](https://github.com/flatpak/flatpak/issues/1447#issuecomment-454618542) which seems more robust.